### PR TITLE
introduce platform stm32mp1

### PR DESCRIPTION
**Pull request by etienne-lms**
Introduce platform stm32mp1 with board stm32mp1-stm32mp157c-ev1 based on stm32mp1 SoC family integrating Arm Cortex-A7 technology. In its default configuration, stm32mp1 OP-TEE core operates in a 256kB secure RAM with pager support enabled.

Platform flavor identifiers for the stm32mp1 do contain a hyphen, as the default stm32mp1-stm32mp157c-ev1. For optee_os to support hyphens inside platform flavor, the P-R includes a change in the core make script.

### CODE REVIEW

File: core/arch/arm/plat-stm32mp1/conf.mk [hunk 1]
```diff
@@ -0,0 +1,34 @@
+PLATFORM_FLAVOR ?= stm32mp157c-ev1
+
+include core/arch/arm/cpu/cortex-a7.mk
+core_arm32-platform-aflags += -mfpu=neon
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/conf.mk]: Needed? Why?

File: core/drivers/stm32_uart.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: BSD-2-Clause
```
Comment (jforissier) [inline core/drivers/stm32_uart.c]: Minor: perhaps this should be introduced in a separate commit

File: core/arch/arm/plat-stm32mp1/stm32mp1_misc.S [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/stm32mp1_misc.S]: Minor: file name could be improved. `reset.S` maybe? In any case `stm32mp1_` is redundant since we're already in `plat-stm32mp1`.

File: core/arch/arm/plat-stm32mp1/scripts/stm32image.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/scripts/stm32image.py]: This script has a number of small formatting issues, could you run `pep8` and see if you can address most of the warnings/errors?

File: core/arch/arm/plat-stm32mp1/link.mk [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+all: $(link-out-dir)/tee-pager_v2.stm32
+cleanfiles += $(link-out-dir)/tee-pager_v2.stm32
+$(link-out-dir)/tee-pager_v2.stm32: $(link-out-dir)/tee-pager_v2.bin
+	$(stm32image_cmd) --source $< --dest $@
+
+all: $(link-out-dir)/tee-pageable_v2.stm32
+cleanfiles += $(link-out-dir)/tee-pageable_v2.stm32
+$(link-out-dir)/tee-pageable_v2.stm32: $(link-out-dir)/tee-pageable_v2.bin
+	@[ `stat -c %s $<` -ne 0 ] || echo "" > $<
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/link.mk]: Looks like a hack ;-) A rule should not modify its prerequisites. How about fixing `stm32image.py` so that it deals with empty files instead?

File: core/core.mk [hunk 6]
```diff
@@ -11,7 +11,7 @@ include mk/config.mk
 include core/arch/$(ARCH)/$(ARCH).mk
 
 PLATFORM_$(PLATFORM) := y
-PLATFORM_FLAVOR_$(PLATFORM_FLAVOR) := y
+PLATFORM_FLAVOR_$(subst -,_,$(PLATFORM_FLAVOR)):= y
```
Comment (jforissier) [inline core/core.mk]: I'm not sure about this patch. As it is, it is a bit incomplete: why allow hyphens in the flavor but not in the platform? Also, the line that splits a composite PLATFORM into PLATFORM and PLATFORM_FLAVOR could be modified to address the limitation you're mentioning in the commit message (split at the first hyphen by default).

However it may be more reasonable to stick to the "no hyphen" practice. We have two instances already, `qemu_virt` and `salvator_h3`.

File: core/arch/arm/plat-stm32mp1/link.mk (hunk 5 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/link.mk]: :) ok

File: core/arch/arm/plat-stm32mp1/scripts/stm32image.py (hunk 4 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/scripts/stm32image.py]: ok. thanks

File: core/arch/arm/plat-stm32mp1/stm32mp1_misc.S (hunk 3 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/stm32mp1_misc.S]: ok :)

File: core/core.mk (hunk 6 above)
Comment (etienne-lms) [reply core/core.mk]: > why allow hyphens in the flavor but not in the platform?

With legacy build scripts, an hyphen in `PLATFORM` is implicitly a separation between a platform and its flavor. Do you want to discard this? And expect all build explicitly set `PLATFORM_FLAVOR` when needed to specify a flavor?

> However it may be more reasonable to stick to the "no hyphen" practice. 

Why not allowing `PLATFORM_FLAVOR` to contain hyphen(s)?

File: core/drivers/stm32_uart.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/stm32_uart.c]: ok

Comment (etienne-lms) [reply core/drivers/stm32_uart.c]: I will split commits for the platform from its uart driver was squashing the changes.

**Comment by etienne-lms** [comment]
comments on platform addressed.

File: core/arch/arm/plat-stm32mp1/scripts/stm32image.py (hunk 4 above)
Comment (jbech-linaro) [reply core/arch/arm/plat-stm32mp1/scripts/stm32image.py]: Please also make sure it works with Python3 (Python 2 have [no maintenance in 2020](https://pythonclock.org/)). So, maybe even better to have `#!/usr/bin/env python3`

File: core/core.mk (hunk 6 above)
Comment (jforissier) [reply core/core.mk]: > With legacy build scripts, an hyphen in PLATFORM is implicitly a separation between a platform and its flavor. Do you want to discard this?

No, I would make it "the **first** hyphen is a separation between (base) platform and its flavor".

> Why not allowing PLATFORM_FLAVOR to contain hyphen(s)?

Well, it is a bit weird to allow hyphens in flavor but not in the base platform, and it makes things a bit complicated. The hyphen cannot stay in the derived name $(PLATFORM_$(PLATFORM_FLAVOR)), so it is simpler to forbid the hyphen everywhere.

Anyway, if you prefer to allow hyphens in the flavor, it's OK, but I think this is the way to do it:

```diff
diff --git a/Makefile b/Makefile
index 7caab01bb..e2578017a 100644
--- a/Makefile
+++ b/Makefile
@@ -30,11 +30,20 @@ include $(CFG_OPTEE_CONFIG)
 endif
 
 # If $(PLATFORM) is defined and contains a hyphen, parse it as
-# $(PLATFORM)-$(PLATFORM_FLAVOR) for convenience
+# $(PLATFORM)-$(PLATFORM_FLAVOR) for convenience (split at first hyphen)
 ifneq (,$(findstring -,$(PLATFORM)))
-ops := $(join PLATFORM PLATFORM_FLAVOR,$(addprefix =,$(subst -, ,$(PLATFORM))))
-$(foreach op,$(ops),$(eval override $(op)))
+_P := $(firstword $(subst -, ,$(PLATFORM)))
+_F := $(subst -,_,$(subst $(_P)-,,$(PLATFORM)))
+ifneq (,$(_F))
+ifneq (,$(PLATFORM_FLAVOR))
+$(error PLATFORM_FLAVOR is set but PLATFORM contains -<flavor>)
+endif
+endif
+$(eval override PLATFORM := $(_P))
+$(eval override PLATFORM_FLAVOR := $(_F))
 endif
+# Any remaining hyphen is converted to an underscore
+override PLATFORM_FLAVOR := $(subst -,_,$(PLATFORM_FLAVOR))
 
 # Make these default for now
 $(call force,ARCH,arm)
```

The differences with your proposal:
- `make stm32mp1-stm32mp157c-ev1` works as expected
- Hyphens are always replaced in the flavor name, not just in the C pre-processor macro $(PLATFORM_$(PLATFORM_FLAVOR)). It's just more consistent.
- `make PLATFORM=foo-bar PLATFORM_FLAVOR=<something>` will report an error because flavor is given twice.

Comment (jenswi-linaro) [reply core/core.mk]: Please keep it simple and skip hyphens in flavors.

Comment (etienne-lms) [reply core/core.mk]: Ok, let's keep it simple.
I will discard my change.

File: core/arch/arm/plat-stm32mp1/scripts/stm32image.py (hunk 4 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/scripts/stm32image.py]: ok

File: core/arch/arm/plat-stm32mp1/conf.mk [hunk 7]
```diff
@@ -1,5 +1,3 @@
-PLATFORM_FLAVOR ?= stm32mp157c-ev1
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/conf.mk]: Minor: since you removed the flavor completely, there is nowhere we can find which board exactly the default build will work on (er... is it even proper English syntax? but I think you get the point).

Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/conf.mk]: ok, i get your point. i will restore.

**Comment by etienne-lms** [comment]
thanks @jforissier for the review.
Squashed, rebased, tag applied.

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
thanks @jenswi-linaro, tag applied.
